pkg/operator/reports.go

```diff
@@ -354,38 +354,68 @@ func getReportPeriod(now time.Time, logger log.FieldLogger, report *metering.Rep
 	return reportPeriod, nil
 }
 
+// handleExpiredReport is responsible for determining if the @report object we're
+// currently processing has reached it's configured retention period. In the case
+// where the Report object hasn't been configured with an expiration, exit early,
+// else attempt to determine if we can safely delete this object without breaking
+// other Report custom resources that rely on this particular Report.
 func (op *Reporting) handleExpiredReport(logger log.FieldLogger, report *metering.Report, now time.Time) error {
-	// check if the Report is being deleted already
+	// check if the Report deletion timestamp has already been set, exit early.
 	if report.DeletionTimestamp != nil {
-		logger.Warnf("report was already marked for deletion")
+		// TODO: should we be requeuing here?
```

Comment:
returning `nil` from this func allows `syncReport` to keep going on the Report object -- has no effect on the handling of the Report, right?

Reply:
Yeah returning nil here would allow `op.syncReport` to continue processing the resource. I think we may want to just re-queue here with the assumption that if the deletion timestamp is not nil, we should assume it's been deleted either by us or the user, but let the informer determine what's the best course of action to take.

Reply:
Actually @pweil- are you aware of any best practices as to when we should be re-queuing resources we're currently processing vs. exiting early and letting the call site determine how to handle the resource from there (assuming there's sufficient logic for the latter case.)

Reply:
Is the more appropriate question to ask "why are we even in this method for a deleted or unexpired report"?  The orchestrator (`syncReport` or `handleReport`) can take care of this for us by calling `isReportExpired` and only calling `handleExpiredReport` when necessary (ie. it's expired and not deleted) or requeing the object, no?

Then you aren't tangling up the orchestration logic with the handler logic as much.  Which incidentally is my answer to your original question: avoid coupling the orchestrator that determines state and next action with the handler that performs a discreet piece of work.

Handlers that defend against bad input are good though but no work to do is success in most cases.

Reply:
Yeah agreed, that sounds more reasonable than how we're currently handing processing resources during upgrades. @bentito I think if we do some light refactoring of these methods based on what Paul's comment, we can also further weed out situations where we don't need to create a deep copy (e.g. are we processing an expired or finished report?) and save the overhead of making that kind of expensive call.

Reply:
I agree, the changes to those methods should save unneeded work.